Add Photon notifications to AMO

kumar303 opened the pull request:
Fixes https://github.com/mozilla/addons-frontend/issues/3706

I needed success notices for the add-to-collection feature so I implemented most of [the Photon notifications](https://design.firefox.com/photon/components/message-bars.html). 

These features have *not* been implemented:
* Generic notices or warnings
* Discopane. I started trying to update this code but it was copy/pasted and has complications due to the flex rules. I can't fit it into this patch.
* Messages are not yet dismissable

<img width="438" alt="screenshot 2018-01-16 11 33 27" src="https://user-images.githubusercontent.com/55398/35003016-1b5a4370-fab1-11e7-9b50-1abd4ee4e2d0.png">
<img width="632" alt="screenshot 2018-01-12 12 34 00" src="https://user-images.githubusercontent.com/55398/34894320-8a278398-f7a6-11e7-838e-2b14fb625d07.png">
<img width="322" alt="screenshot 2018-01-16 11 34 29" src="https://user-images.githubusercontent.com/55398/35003066-40d08f1a-fab1-11e7-9be8-adb82bb3f239.png">


Here is a success notice:

<img width="530" alt="screenshot 2018-01-12 08 52 50" src="https://user-images.githubusercontent.com/55398/34894338-9ca5da06-f7a6-11e7-85eb-069f49ec7cf8.png">

Here's an error on the review page:

<img width="348" alt="screenshot 2018-01-16 11 29 51" src="https://user-images.githubusercontent.com/55398/35002867-a048371e-fab0-11e7-9d25-3520d1e5eb21.png">

Here's how it looks with an action button:

<img width="322" alt="screenshot 2018-01-16 11 25 51" src="https://user-images.githubusercontent.com/55398/35002678-0f38ddb4-fab0-11e7-9f15-fedec6787f27.png">
<img width="651" alt="screenshot 2018-01-12 09 59 16" src="https://user-images.githubusercontent.com/55398/34894376-c3cc74a0-f7a6-11e7-9a04-4aa047aad379.png">
<img width="428" alt="screenshot 2018-01-12 09 59 35" src="https://user-images.githubusercontent.com/55398/34894377-c3f874e2-f7a6-11e7-983a-0f2d38912ce4.png">
<img width="351" alt="screenshot 2018-01-16 11 27 14" src="https://user-images.githubusercontent.com/55398/35002746-42186718-fab0-11e7-921e-78dd6e016e5d.png">
<img width="623" alt="screenshot 2018-01-12 10 01 11" src="https://user-images.githubusercontent.com/55398/34894379-c42fa8cc-f7a6-11e7-835b-330a2df7c238.png">

willdurand commented:
😍

tofumatt commented:
It looks great! 👍 I'll probably get to the review later tonight or tomorrow, but thanks for this!

tofumatt commented on src/ui/components/Notice/index.js:
```diff
@@ -0,0 +1,67 @@
+/* @flow */
+import makeClassName from 'classnames';
+import * as React from 'react';
```
Why not just `import React from 'react';`?

kumar303 replied:
The [package](https://reactjs.org/blog/2017/09/26/react-v16.0.html#packaging) changed in React 16 so importing `*` makes more things work as expected.

tofumatt commented on src/ui/components/Notice/styles.scss:
```diff
@@ -0,0 +1,97 @@
+  font-size: $font-size-s;
+  grid-gap: 4px;
+  grid-template-columns: 20px auto;
+  line-height: 15px;
```
`line-height` [should be in unitless measurements](https://css-tricks.com/almanac/properties/l/line-height/#article-header-id-0); I'm guessing this is probably around `1.2` but I'm not sure, you'll have to trial+error it 🙃

tofumatt commented on src/ui/components/Notice/styles.scss:
```diff
@@ -0,0 +1,97 @@
+  border-radius: $border-radius-s;
+  display: grid;
+  font-size: $font-size-s;
+  grid-gap: 4px;
```
Most of our base units are `6px` multiples. Should this really be `4px`? If it totally looks right compared to mocks that's okay, but I'd assume 6 is preferred.

kumar303 replied:
The [spec](https://design.firefox.com/photon/components/message-bars.html) says it should be 4px.

tofumatt commented on src/ui/components/Notice/styles.scss:
```diff
@@ -0,0 +1,97 @@
+  grid-gap: 4px;
+  grid-template-columns: 20px auto;
+  line-height: 15px;
+  min-height: 20px;
```
Same throughout re: `6px` base multiple; these all seem to be multiples of `4` which isn't common in Add-ons or Photon and is in fact a new multiple.

tofumatt commented on src/ui/components/Notice/styles.scss:
```diff
@@ -0,0 +1,97 @@
+  &.Button--small {
+    border: none;
+    font-size: $font-size-s;
+    line-height: 15px;
```
This should be unitless too.

tofumatt commented on src/ui/components/Notice/styles.scss:
```diff
@@ -0,0 +1,97 @@
+    line-height: 15px;
+    padding: 6px;
+    text-shadow: none;
+    width: max-content;
```
This seems to be unsupported in Edge: https://caniuse.com/#search=max-content

That's not the end of the world IMO but I'd rather us be compatible. Is there any way to replicate what you're doing here and have it work with Edge?

tofumatt replied:
Ah, okay, that's fine then. 👍

tofumatt commented on tests/unit/ui/components/TestNotice.js:
```diff
@@ -0,0 +1,58 @@
+    const root = render();
+
+    expect(root).not.toHaveClassName('.Notice-with-action');
+    // There should only be one column besides the icon.
```
This is an uber-nitpick but "besides the icon" is a bit confusing to read–I couldn't tell at first if it meant "other than the icon" or "next to the icon". I think "one column other than the icon column" would clarify things.

kumar303 commented:
Thanks for the review! I'm back from the holiday so I'll start working on this after breakfast. Did you look at [the spec](https://design.firefox.com/photon/components/message-bars.html)? It's pretty nice because it has pixel dimensions and it specifies button placement, etc. Some changes you requested go against that spec. If you would like to do that, we should probably get approval from someone on the Photon project.

tofumatt commented:
Oh, I see. I hadn’t seen that part of the spec.

I definitely don’t like that 😅 But you’re right: it’s part of the spec. I’ll bring it up with people but you’re right: you should follow the spec for now.

kumar303 commented:
I updated the multi-line message based on a "spec" that @pwalm just gave me 😃 I think this looks better (I adjusted the screenshots). 
![img_1255](https://user-images.githubusercontent.com/55398/35003375-458918a0-fab2-11e7-8c3a-2d2cd6b5bf74.JPG)

kumar303 commented:
Tests + lint all pass for me locally
